pkg/server/backend_manager.go

```diff
@@ -95,23 +95,25 @@ func NewDefaultBackendManager() *DefaultBackendManager {
 }
 
 // AddBackend adds a backend.
-func (s *DefaultBackendManager) AddBackend(agentID string, conn agent.AgentService_ConnectServer) {
+func (s *DefaultBackendManager) AddBackend(agentID string, conn agent.AgentService_ConnectServer) Backend {
 	klog.Infof("register Backend %v for agentID %s", conn, agentID)
 	s.mu.Lock()
 	defer s.mu.Unlock()
 	_, ok := s.backends[agentID]
+	addedBackend := newBackend(conn)
 	if ok {
 		for _, v := range s.backends[agentID] {
 			if v.conn == conn {
 				klog.Warningf("this should not happen. Adding existing connection %v for agentID %s", conn, agentID)
-				return
+				return nil
```

Comment:
This is dangerous and could lead to some bizarre and hard to track down bugs. We will take this nil and use it as the key in frontends to identify the backend. We then won't properly close down the frontend when its actual backend connection break. If we ever call getFrontendsForBackendConn with a backend of nil we could have multiple unrelated frontends all getting closed. Guessing we should just return v.

Reply:
Good point, fixed.